libs/telemetry/src/main/java/org/opensearch/telemetry/tracing/DefaultTracer.java

```diff
@@ -77,15 +82,51 @@ public SpanContext getCurrentSpan() {
         return (currentSpan == null) ? null : new SpanContext(currentSpan);
     }
 
-    private void endSpan(Span span) {
-        if (span != null) {
-            span.endSpan();
-            setCurrentSpanInContext(span.getParentSpan());
-        }
+    @Override
+    public ScopedSpan startScopedSpan(SpanCreationContext spanCreationContext) {
+        Span span = startSpan(spanCreationContext);
+        SpanScope spanScope = createSpanScope(span);
+        return new DefaultScopedSpan(
+            span,
+            spanScope,
+            (spanToBeClosed, scopeSpanToBeClosed) -> endScopedSpan(spanToBeClosed, scopeSpanToBeClosed)
+        );
+    }
+
+    @Override
+    public ScopedSpan startScopedSpan(SpanCreationContext spanCreationContext, SpanContext parentSpan) {
+        Span span = startSpan(spanCreationContext.getSpanName(), parentSpan, spanCreationContext.getAttributes());
+        SpanScope spanScope = createSpanScope(span);
+        return new DefaultScopedSpan(
+            span,
+            spanScope,
+            (spanToBeClosed, scopeSpanToBeClosed) -> endScopedSpan(spanToBeClosed, scopeSpanToBeClosed)
+        );
+    }
+
+    @Override
+    public SpanScope createSpanScope(Span span) {
+        return DefaultSpanScope.create(span);
+    }
+
+    private void endScopedSpan(Span span, SpanScope spanScope) {
+        span.endSpan();
+        spanScope.close();
     }
 
     private Span createSpan(String spanName, Span parentSpan, Attributes attributes) {
-        return tracingTelemetry.createSpan(spanName, parentSpan, attributes);
+        return tracingTelemetry.createSpan(spanName, parentSpan, attributes, span -> onSpanEnd(span, parentSpan));
+    }
+
+    private void onSpanEnd(Span span, Span parentSpan) {
```

Comment:
As we discussed, let's make sure that `getCurrentSpanInContext()` == this span so we won't be screwing the context up.

Reply:
We can handle it like this 

```
if(getCurrentSpanInternal() != span) {
    span.setError(new IllegalStateException(String.format("Span not closed properly, expected current span on closure [%s] but found [%s]", span, getCurrentSpanInternal()));
}
```

Reply:
I think that would be asking for a problem right away, example here:

```
Span span = tracer.createSpan(...); // T1
try (tracer.withSpanInScope()) {
    someAsyncOperation(...).whenComplete(() -> span.end()); // T2
}
```

In this case, the `span.end()` that ends on T2 would unlikely have any span in the storage.

Reply:
Yes, this should be the case. But here do you want me to add assertion to make sure that the else case should always be `assert getCurrentSpanInternal () == span`

Reply:
> But here do you want me to add assertion to make sure that the else case should always be `assert getCurrentSpanInternal () == span`

No, I think it is just sufficient to check if `getCurrentSpanInternal () == span` than restore the parent back, but it is not - do nothing.